File: TP4/Q04/Show.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.PrintWriter;
import java.io.IOException;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.*;

public class Show {
    private String SHOW_ID;
    private String TYPE;
    private String TITLE;
    private String DIRECTOR;
    private String[] CAST;
    private String COUNTRY;
    private Date DATE_ADDED;
    private int RELEASE_YEAR;
    private String RATING;
    private String DURATION;
    private String[] LISTED_IN;

    public Show() {
        this.SHOW_ID = "NaN";
        this.TYPE = "NaN";
        this.TITLE = "NaN";
        this.DIRECTOR = "NaN";
        this.CAST = new String[]{"NaN"};
        this.COUNTRY = "NaN";
        this.DATE_ADDED = null;
        this.RELEASE_YEAR = -1;
        this.RATING = "NaN";
        this.DURATION = "NaN";
        this.LISTED_IN = new String[]{"NaN"};
    }

    public Show(String SHOW_ID, String TYPE, String TITLE, String DIRECTOR, String[] CAST, String COUNTRY,
                Date DATE_ADDED, int RELEASE_YEAR, String RATING, String DURATION, String[] LISTED_IN) {
        this.SHOW_ID = SHOW_ID;
        this.TYPE = TYPE;
        this.TITLE = TITLE;
        this.DIRECTOR = (DIRECTOR.isEmpty()) ? "NaN" : DIRECTOR;
        this.CAST = (CAST.length == 0 || CAST[0].isEmpty()) ? new String[]{"NaN"} : CAST;
        this.COUNTRY = (COUNTRY.isEmpty()) ? "NaN" : COUNTRY;
        this.DATE_ADDED = DATE_ADDED;
        this.RELEASE_YEAR = RELEASE_YEAR;
        this.RATING = (RATING.isEmpty()) ? "NaN" : RATING;
        this.DURATION = (DURATION.isEmpty()) ? "NaN" : DURATION;
        this.LISTED_IN = (LISTED_IN.length == 0 || LISTED_IN[0].isEmpty()) ? new String[]{"NaN"} : LISTED_IN;
    }

    public String getTITLE() { return TITLE; }
    public String getDIRECTOR() { return DIRECTOR; }
    public String getSHOW_ID() { return SHOW_ID; }
    public String getTYPE() { return TYPE; }
    public String getCOUNTRY() { return COUNTRY; }
    public Date getDATE_ADDED() { return DATE_ADDED; }
    public int getRELEASE_YEAR() { return RELEASE_YEAR; }
    public String getRATING() { return RATING; }
    public String getDURATION() { return DURATION; }
    public String[] getCAST() { return CAST; }
    public String[] getLISTED_IN() { return LISTED_IN; }

    public static Show buscarPorID(String idBuscado, String caminhoArquivo) {
        try (BufferedReader br = new BufferedReader(new FileReader(caminhoArquivo))) {
            String linha = br.readLine(); // pular header
            while ((linha = br.readLine()) != null) {
                String[] dados = linha.split(",(?=(?:[^\"]*\"[^\"]*\")*[^\"]*$)", -1);
                if (dados.length < 12) continue;

                String show_id = dados[0].trim();
                if (!show_id.equalsIgnoreCase(idBuscado)) continue;

                String type = dados[1].trim().replaceAll("\"", "");
                String title = dados[2].trim().replaceAll("\"", "");
                String director = dados[3].trim().replaceAll("\"", "");
                String[] cast = (dados[4].trim().isEmpty()) ? new String[]{"NaN"} : dados[4].trim().split(", ");
                String country = dados[5].trim();

                Date dateAdded = null;
                if (!dados[6].trim().isEmpty()) {
                    try {
                        String dataLimpa = dados[6].trim().replaceAll("\"", "");
                        SimpleDateFormat formato = new SimpleDateFormat("MMMM d, yyyy", Locale.ENGLISH);
                        dateAdded = formato.parse(dataLimpa);
                    } catch (ParseException e) {
                        dateAdded = null;
                    }
                }

                int releaseYear = -1;
                if (!dados[7].trim().isEmpty()) {
                    try {
                        releaseYear = Integer.parseInt(dados[7].trim().replaceAll("\"", ""));
                    } catch (NumberFormatException e) {
                        releaseYear = -1;
                    }
                }

                String rating = (dados[8].trim().isEmpty()) ? "NaN" : dados[8].trim().replaceAll("\"", "");
                String duration = dados[9].trim();
                String[] listedIn = (dados[10].trim().isEmpty()) ? new String[]{"NaN"} : dados[10].trim().split(", ");

                return new Show(show_id, type, title, director, cast, country, dateAdded, releaseYear, rating, duration, listedIn);
            }
        } catch (IOException e) {
            System.out.println("Erro ao ler o arquivo: " + e.getMessage());
        }
        return null;
    }

    private enum Cor { RED, BLACK }

    private static class No {
        Show elemento;
        No esq, dir;
        Cor cor;

        public No(Show elem) {
            this.elemento = elem;
            this.cor = Cor.RED;
        }
    }

    public static class ArvoreAlvinegra {
        private No raiz;
        private int comparacoes = 0;

        private boolean isRed(No no) {
            return no != null && no.cor == Cor.RED;
        }

        private No rotacionarEsq(No h) {
            No x = h.dir;
            h.dir = x.esq;
            x.esq = h;
            x.cor = h.cor;
            h.cor = Cor.RED;
            return x;
        }

        private No rotacionarDir(No h) {
            No x = h.esq;
            h.esq = x.dir;
            x.dir = h;
            x.cor = h.cor;
            h.cor = Cor.RED;
            return x;
        }

        private void trocarCor(No h) {
            h.cor = (h.cor == Cor.RED) ? Cor.BLACK : Cor.RED;
            if (h.esq != null) h.esq.cor = (h.esq.cor == Cor.RED) ? Cor.BLACK : Cor.RED;
            if (h.dir != null) h.dir.cor = (h.dir.cor == Cor.RED) ? Cor.BLACK : Cor.RED;
        }

        public void inserir(Show show) {
            raiz = inserirRec(raiz, show);
            raiz.cor = Cor.BLACK;
        }

        private No inserirRec(No h, Show show) {
            if (h == null) return new No(show);

            comparacoes++;
            int cmp = show.getTITLE().compareTo(h.elemento.getTITLE());
            if (cmp < 0) {
                h.esq = inserirRec(h.esq, show);
            } else if (cmp > 0) {
                h.dir = inserirRec(h.dir, show);
            } 

            if (isRed(h.dir) && !isRed(h.esq)) h = rotacionarEsq(h);
            if (isRed(h.esq) && isRed(h.esq.esq)) h = rotacionarDir(h);
            if (isRed(h.esq) && isRed(h.dir)) trocarCor(h);

            return h;
        }

        public boolean pesquisar(String titulo) {
            System.out.print("=>raiz ");
            No p = raiz;
            while (p != null) {
                comparacoes++;
                if (titulo.equals(p.elemento.getTITLE())) {
                    return true;
                } else if (titulo.compareTo(p.elemento.getTITLE()) < 0) {
                    System.out.print("esq ");
                    p = p.esq;
                } else {
                    System.out.print("dir ");
                    p = p.dir;
                }
            }
            return false;
        }

        public int getComparacoes() {
            return comparacoes;
        }
    }

    public static void main(String[] args) throws Exception {
        Scanner sc = new Scanner(System.in);
        String caminhoArquivo = "/tmp/disneyplus.csv";
        ArvoreAlvinegra arvore = new ArvoreAlvinegra();

        while (true) {
            String entrada = sc.nextLine().trim();
            if (entrada.equalsIgnoreCase("FIM")) break;
            Show show = buscarPorID(entrada, caminhoArquivo);
            if (show != null) arvore.inserir(show);
        }

        long inicio = System.nanoTime();
        while (sc.hasNext()) {
            String tituloPesquisa = sc.nextLine().trim();
            if (tituloPesquisa.equalsIgnoreCase("FIM")) break;
            boolean encontrado = arvore.pesquisar(tituloPesquisa);
            System.out.println(encontrado ? "SIM" : "NAO");
        }
        long fim = System.nanoTime();

        double tempoExecucao = (fim - inicio) / 1e6;
        String matricula = "854155";

        try (PrintWriter writer = new PrintWriter(new FileWriter(matricula + "_avinegra.txt"))) {
            writer.printf("%s\t%.3f\t%d\n", matricula, tempoExecucao, arvore.getComparacoes());
        } catch (IOException e) {
            System.out.println("Erro ao escrever o log: " + e.getMessage());
        }

        sc.close();
    }
}

File: TP4/Q03/Show.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

#define MAX_TAM 1000
#define CSV_PATH "/tmp/disneyplus.csv"

typedef struct {
    char show_id[20];
    char type[30];
    char title[200];
} Show;

typedef struct Node {
    Show show;
    int altura;
    struct Node* esq;
    struct Node* dir;
} Node;

int comparacoes = 0;

int max(int a, int b) {
    return (a > b) ? a : b;
}

int altura(Node* no) {
    return no == NULL ? 0 : no->altura;
}

int balanceamento(Node* no) {
    return (no == NULL) ? 0 : altura(no->esq) - altura(no->dir);
}

Node* rotacionarDir(Node* y) {
    Node* x = y->esq;
    Node* T2 = x->dir;
    x->dir = y;
    y->esq = T2;

    y->altura = max(altura(y->esq), altura(y->dir)) + 1;
    x->altura = max(altura(x->esq), altura(x->dir)) + 1;
    return x;
}

Node* rotacionarEsq(Node* x) {
    Node* y = x->dir;
    Node* T2 = y->esq;
    y->esq = x;
    x->dir = T2;

    x->altura = max(altura(x->esq), altura(x->dir)) + 1;
    y->altura = max(altura(y->esq), altura(y->dir)) + 1;
    return y;
}

Node* inserir(Node* raiz, Show show) {
    if (raiz == NULL) {
        Node* novo = (Node*)malloc(sizeof(Node));
        novo->show = show;
        novo->esq = novo->dir = NULL;
        novo->altura = 1;
        return novo;
    }

    comparacoes++;
    if (strcmp(show.title, raiz->show.title) < 0) {
        raiz->esq = inserir(raiz->esq, show);
    } else if (strcmp(show.title, raiz->show.title) > 0) {
        raiz->dir = inserir(raiz->dir, show);
    } else {
        return raiz;
    }

    raiz->altura = 1 + max(altura(raiz->esq), altura(raiz->dir));

    int balance = balanceamento(raiz);

    if (balance > 1 && strcmp(show.title, raiz->esq->show.title) < 0)
        return rotacionarDir(raiz);
    if (balance < -1 && strcmp(show.title, raiz->dir->show.title) > 0)
        return rotacionarEsq(raiz);
    if (balance > 1 && strcmp(show.title, raiz->esq->show.title) > 0) {
        raiz->esq = rotacionarEsq(raiz->esq);
        return rotacionarDir(raiz);
    }
    if (balance < -1 && strcmp(show.title, raiz->dir->show.title) < 0) {
        raiz->dir = rotacionarDir(raiz->dir);
        return rotacionarEsq(raiz);
    }

    return raiz;
}

int pesquisar(Node* raiz, char* titulo) {
    printf("=>raiz ");
    while (raiz != NULL) {
        comparacoes++;
        int cmp = strcmp(titulo, raiz->show.title);
        if (cmp == 0) {
            return 1;
        } else if (cmp < 0) {
            printf("esq ");
            raiz = raiz->esq;
        } else {
            printf("dir ");
            raiz = raiz->dir;
        }
    }
    return 0;
}

Show buscarPorID(char* idBuscado) {
    FILE* fp = fopen(CSV_PATH, "r");
    char linha[1024];
    Show s;
    strcpy(s.title, ""); 

    if (fp == NULL) {
        printf("Erro ao abrir arquivo.\n");
        return s;
    }

    fgets(linha, sizeof(linha), fp); 

    while (fgets(linha, sizeof(linha), fp) != NULL) {
        char* token = strtok(linha, ",");
        if (token && strcmp(token, idBuscado) == 0) {
            strcpy(s.show_id, token);
            token = strtok(NULL, ","); 
            strcpy(s.type, token ? token : "NaN");
            token = strtok(NULL, ","); 
            strcpy(s.title, token ? token : "NaN");
            break;
        }
    }

    fclose(fp);
    return s;
}

int main() {
    char entrada[100];
    Node* raiz = NULL;
    clock_t inicio, fim;

    while (scanf(" %s", entrada), strcmp(entrada, "FIM") != 0) {
        Show s = buscarPorID(entrada);
        if (strcmp(s.title, "") != 0) {
            raiz = inserir(raiz, s);
        }
    }

    inicio = clock();
    while (scanf(" %[^\n]", entrada), strcmp(entrada, "FIM") != 0) {
        int encontrado = pesquisar(raiz, entrada);
        printf(encontrado ? "SIM\n" : "NAO\n");
    }
    fim = clock();

    double tempo = (double)(fim - inicio) * 1000.0 / CLOCKS_PER_SEC;
    FILE* log = fopen("854155_avl.txt", "w");
    if (log != NULL) {
        fprintf(log, "854155\t%.3lf\t%d\n", tempo, comparacoes);
        fclose(log);
    }

    return 0;
}
